brave-instrumentation/aws-java-sdk-core/src/main/java/brave/instrumentation/aws/TracingRequestHandler.java

```diff
@@ -36,79 +38,77 @@
  */
 public final class TracingRequestHandler extends RequestHandler2 {
 
-  public static TracingRequestHandler create(HttpTracing httpTracing) {
-    return new TracingRequestHandler(httpTracing);
-  }
-
-  public static TracingRequestHandler create(Tracing tracing) {
-    return new TracingRequestHandler(HttpTracing.create(tracing));
-  }
-
+  static final HandlerContextKey<Span> APPLICATION_SPAN = new HandlerContextKey<>("APPLICATION_SPAN");
   static final HandlerContextKey<Span> SPAN = new HandlerContextKey<>(Span.class.getCanonicalName());
-  static final HandlerContextKey<TracingRequestHandler> TRACING_REQUEST_HANDLER_CONTEXT_KEY = new HandlerContextKey<>(TracingRequestHandler.class.getCanonicalName());
 
   static final HttpClientAdapter<Request<?>, Response<?>> ADAPTER = new HttpAdapter();
 
-  static final Propagation.Setter<Request<?>, String> SETTER = new Propagation.Setter<Request<?>, String>() {
-    @Override public void put(Request<?> carrier, String key, String value) {
-      carrier.addHeader(key, value);
-    }
-  };
+  static final Propagation.Setter<Request<?>, String> SETTER = (carrier, key, value) -> carrier.addHeader(key, value);
 
+  final Tracer tracer;
   final HttpClientHandler<Request<?>, Response<?>> handler;
   final TraceContext.Injector<Request<?>> injector;
 
   TracingRequestHandler(HttpTracing httpTracing) {
+    tracer = httpTracing.tracing().tracer();
     handler = HttpClientHandler.create(httpTracing, ADAPTER);
     injector = httpTracing.tracing().propagation().injector(SETTER);
   }
 
-  @Override public final void beforeRequest(Request<?> request) {
-    Span span = handler.handleSend(injector, request);
-    span.remoteEndpoint(
-        Endpoint.newBuilder().serviceName(request.getServiceName()).build());
-    span.tag("aws.service_name", request.getServiceName());
-    span.tag("aws.operation", getAwsOperationFromRequest(request));
-    request.addHandlerContext(SPAN, span);
-    request.addHandlerContext(TRACING_REQUEST_HANDLER_CONTEXT_KEY, this);
+  @Override
+  public AmazonWebServiceRequest beforeExecution(AmazonWebServiceRequest request) {
+    Span span = tracer.nextSpan();
+    span.start();
+    request.addHandlerContext(APPLICATION_SPAN, span);
+
+    return request;
+  }
+
+  @Override
+  public void beforeAttempt(HandlerBeforeAttemptContext context) {
+    Span applicationSpan = context.getRequest().getHandlerContext(APPLICATION_SPAN);
+    applicationSpan.name(context.getRequest().getServiceName() + "." + getAwsOperationFromRequest(context.getRequest()));
```

Comment:
since you already have a service name tag, do you need to concatenate it here? usually we don't put dots in span names (well maybe we do but I forget if so..)